### PR TITLE
Add support for setting shape inference function on fused nodes

### PR DESCRIPTION
**Pull request by pranav-prakash**
**Description**: See #7006. 

The ability to manually set a shape inference function for function-based nodes is useful for execution providers who perform node-fusion during graph partitioning. While normally shape info is computed before partitioning and hence is preserved by node fusion, in the case of graph transforms that operate _after_ partitioning (level 3 and above, e.g. NHWC graph transformations) shape info can be reset or new nodeArgs created. In such cases, rather than manually setting shape info in every graph transform that deals with fused nodes, it can be cleaner to just define the shape inference function upfront.

### CODE REVIEW

**Comment by snnn** [comment]
Hi @askhade, could you please help?

File: onnxruntime/test/framework/inference_session_test.cc [hunk 1]
```diff
@@ -1224,6 +1244,88 @@ TEST(ExecutionProviderTest, FunctionTest) {
...
+  ASSERT_TRUE(status.IsOK());
+  status = session.Initialize();
+  ASSERT_TRUE(status.IsOK());
+
+  const Graph& fused_graph = session.GetModel().MainGraph();
+  ASSERT_TRUE(fused_graph.NumberOfNodes() == 1);
+  for (const Node& node : fused_graph.Nodes()) {
+    ASSERT_TRUE(node.NodeType() == Node::Type::Fused);
+    ASSERT_TRUE(node.Op()->has_type_and_shape_inference_function());
+  }
```
Comment (skottmckay) [inline onnxruntime/test/framework/inference_session_test.cc:1326]: I would have expected the test to check the inferred output shape after Graph::Resolve is called. actually executing the model seems optional, and doesn't necessarily prove the type/shape inferencing worked as we handle unknown output shapes. 

should be able to look at the output definitions for the fused node to check the shape is set and is as expected (using Node::OutputDefs() and NodeArg::Shape()).

File: include/onnxruntime/core/graph/indexed_sub_graph.h [hunk 2]
```diff
@@ -36,6 +36,9 @@ struct IndexedSubGraph {
     NodeAttributes attributes;         ///< Attributes of customized SubGraph/FunctionProto.
 
     std::string doc_string;  ///< Doc string of customized SubGraph/FunctionProto.
+    
+    /** Type and shape inference function that should be defined for the fused node */
+    std::function<void (ONNX_NAMESPACE::InferenceContext&)> inference_function;
```
Comment (skottmckay) [inline include/onnxruntime/core/graph/indexed_sub_graph.h:41]: nit: comment is misleading. 'should be defined' -> 'can optionally be defined' 

'inference_function' is a bit vague as well. can we make it 'type_and_shape_inference_function'?

File: onnxruntime/test/framework/inference_session_test.cc [hunk 3]
```diff
@@ -62,6 +62,15 @@ struct KernelRegistryAndStatus {
...
 namespace onnxruntime {
+
+class TestingInferenceSession : public InferenceSession {
+public:
+  using InferenceSession::InferenceSession;
+  const Model& GetModel() {
+    return *model_;
+  }
+};
+
```
Comment (skottmckay) [inline onnxruntime/test/framework/inference_session_test.cc:73]: Please use InferenceSessionWrapper

File: include/onnxruntime/core/graph/indexed_sub_graph.h (hunk 2 above)
Comment (pranav-prakash) [reply include/onnxruntime/core/graph/indexed_sub_graph.h]: Fixed

File: onnxruntime/test/framework/inference_session_test.cc (hunk 3 above)
Comment (pranav-prakash) [reply onnxruntime/test/framework/inference_session_test.cc]: Done, thanks for the pointer (didn't realize this existed). We need access to the mutable graph though (in order to call `.resolve()` which clears out the inferred shapes) so I added an additional `GetMutableGraph` method to `InferenceSessionWrapper`.

File: onnxruntime/test/framework/inference_session_test.cc (hunk 1 above)
Comment (pranav-prakash) [reply onnxruntime/test/framework/inference_session_test.cc:1291]: Done. The test now loads the graph with the FusedExecutionProvider enabled, calls `.Resolve()` to perform a fresh round of shape-inference, then verifies that the shape is as expected.

**Comment by pranav-prakash** [comment]
@skottmckay The CI for minimal ORT seems to fail (despite the regular CPU pipeline succeeding). 
```
/onnxruntime_src/include/onnxruntime/core/graph/indexed_sub_graph.h:41:41: error: ‘InferenceContext’ is not a member of ‘onnx’
   41 |     std::function<void (ONNX_NAMESPACE::InferenceContext&)> type_and_shape_inference_function;

```

Should I need to surround this by an `#ifdef` (I'm not too familiar with the minimal ORT builds).

**Comment by skottmckay** [comment]
> @skottmckay The CI for minimal ORT seems to fail (despite the regular CPU pipeline succeeding).
> 
> ```
> /onnxruntime_src/include/onnxruntime/core/graph/indexed_sub_graph.h:41:41: error: ‘InferenceContext’ is not a member of ‘onnx’
>    41 |     std::function<void (ONNX_NAMESPACE::InferenceContext&)> type_and_shape_inference_function;
> ```
> 
> Should I need to surround this by an `#ifdef` (I'm not too familiar with the minimal ORT builds).

Yes please. Put it inside `#if !defined(ORT_MINIMAL_BUILD)`

We have an extremely limited dependency on ONNX in the minimal build in order to minimize the binary size. The type/shape info is saved in the ORT format model used by the minimal build and not recalculated at runtime.

**Comment by pranav-prakash** [comment]
Fixed, thanks for the info! Could you please re-run CI?

File: onnxruntime/core/graph/function.cc [hunk 4]
```diff
@@ -152,6 +152,11 @@ static std::unique_ptr<ONNX_NAMESPACE::OpSchema> CreateSchema(const Graph& graph
   op_schema->SetDomain(meta_def->domain);
   op_schema->SetDoc(meta_def->doc_string);
   op_schema->SinceVersion(meta_def->since_version);
+
+  if (meta_def->type_and_shape_inference_function) {
```
Comment (skottmckay) [inline onnxruntime/core/graph/function.cc:156]: Probably need an ifdef around this usage as well

Comment (pranav-prakash) [reply onnxruntime/core/graph/function.cc:156]: Seems to pass the minimal CI without it; I guess `function.cc` and the unit test are never included in the minimal build?

File: onnxruntime/test/util/include/inference_session_wrapper.h [hunk 5]
```diff
@@ -20,6 +20,10 @@ class InferenceSessionWrapper : public InferenceSession {
     return model_->MainGraph();
   }
 
+  Graph& GetMutableGraph() const {
```
Comment (skottmckay) [inline onnxruntime/test/util/include/inference_session_wrapper.h:23]: I don't think you need to expose a mutable Graph. Resolve is called as part of model loading, so it shouldn't be necessary to call it again.

Comment (pranav-prakash) [reply onnxruntime/test/util/include/inference_session_wrapper.h:23]: The call to `resolve()` is to make sure that we recompute the shape inference data after the EP does the fusing. Unless I'm misunderstanding, in the model loading process all shape inference data will be computed _before_ the EP does the fusing (unless there are level 3 graph transforms which are run subsequently) – this was the whole motivating reason for this PR. So if we don't manually clear out the shape inference info, then we wouldn't be able to distinguish between shape inference performed by the fused node and shape inference info that was present prior to fusing.

Comment (skottmckay) [reply onnxruntime/test/util/include/inference_session_wrapper.h:23]: Graph::Resolve is called if the partitioning changes the Graph. If you need to do a manual call to Graph::Resolve in the test to make it work, that would imply it wouldn't work in normal usage.

Comment (pranav-prakash) [reply onnxruntime/test/util/include/inference_session_wrapper.h:23]: Under normal usage wouldn't the call to `graph::resolve` implicitly be done by any level-3 transformer that had been registered.

Comment (skottmckay) [reply onnxruntime/test/util/include/inference_session_wrapper.h:23]: Resolve could be called after a level 3 transformer changed the graph, or after there was a fusion during partitioning that changed the graph. Either way, if something changes the graph, a call to Graph::Resolve should happen prior to InferenceSession::Initialize completing. 

To validate that Resolve has been called, Graph::GraphResolveNeeded can be used (and a call to Graph::Resolve will do nothing unless GraphResolveNeeded returns true).

Will the test detect if the shape inferencing function is not called though? The shape info from prior to partitioning will still be present for the 'M' output won't it? If so, you could add a way for the FuseExecutionProvider to track that (no mutable Graph needed, assuming the Graph::Resolve is happening as expected), or you could manually remove the shape info for the 'M' output (would need to call NodeArg::ClearShape as well as Graph::SetResolveNeeded).

Comment (pranav-prakash) [reply onnxruntime/test/util/include/inference_session_wrapper.h:23]: Ah you're right, thanks for the info – I was under the impression that a call to `Resolve()` would _always_ recompute _all_ shape-info, but as you mentioned this is not the case.

> or you could manually remove the shape info
I've updated the test to do so

**Review by skottmckay — APPROVED** [review]

**Pull request merged**